### PR TITLE
Override split level in miniaod

### PR DESCRIPTION
**Pull request by arizzi**
This PR implements a dedicated list of what to split/not-to-split for MiniAOD.
Different branches can be optimal with full split or no-split depending on the number of items per event and/or the number of empty (zeroed) datamembers (sub-branches).

The list is obtained simply comparing what takes more or less space running with full split vs no-split.
Savings of 2.6kb/ev on ttbar with PU (out of ~45kb/ev) are expected

### CODE REVIEW

**Comment by arizzi** [comment]
@gpetruc

**Comment by perrotta** [comment]
please test

**Comment by arizzi** [comment]
http://arizzi.web.cern.ch/arizzi/miniaod/split-custom.html
vs
http://arizzi.web.cern.ch/arizzi/miniaod/default-split.html

File: Configuration/Applications/python/ConfigBuilder.py [hunk 1]
```diff
@@ -642,9 +642,8 @@ def doNotInlineEventContent(instance,label = "cms.untracked.vstring(process."+th
                         output.dataset.filterName = cms.untracked.string('StreamALCACombined')
 
                 if "MINIAOD" in streamType:
-                    output.dropMetaData = cms.untracked.string('ALL')
-                    output.fastCloning= cms.untracked.bool(False)
-                    output.overrideInputFileSplitLevels = cms.untracked.bool(True)                      
+                    from PhysicsTools.PatAlgos.slimming.miniAOD_tools import miniAOD_customizeOutput
+		    miniAOD_customizeOutput(output)
```
Comment (slava77) [inline Configuration/Applications/python/ConfigBuilder.py]: is it another case of inconsistent indentation?

File: PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py [hunk 2]
```diff
@@ -94,3 +94,50 @@
         'keep GenRunInfoProduct_*_*_*',
         'keep L1GtTriggerMenuLite_l1GtTriggerMenuLite__*',
 ]
+
+MiniAODOverrideBranchesSplitLevel = cms.untracked.VPSet( [
+cms.untracked.PSet(branch = cms.untracked.string("patMuons_slimmedMuons__*"),splitLevel=cms.untracked.int32(0)),
+cms.untracked.PSet(branch = cms.untracked.string("patElectrons_slimmedElectrons__*"),splitLevel=cms.untracked.int32(0)),
+cms.untracked.PSet(branch = cms.untracked.string("patTaus_slimmedTaus__*"),splitLevel=cms.untracked.int32(0)),
```
Comment (slava77) [inline PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py]: is the list of branches here substantially different from what's in ```MicroEventContent``` ?
Are there any branches that have to stay in the full split or is the real intent of the changes here to make everything in miniAOD to be splitLevel 0?

It looks like most of the miniAOD branches in this PR are set to use the no-split mode.
It seems more practical to use the output module splitLevel parameter to set the default and then override a few branches to have the full split (like packedCandidates),

**Comment by arizzi** [comment]
Yes, there are branches that have to stay in full split, and in fact we got better performance with everything in full split than with everything in no-split (that's the reason we use full split in first place!)

Concerning what is "cheaper" to do... there are 40 branches out of 80 to keep in full split and 40 to have un-split... so there is no cheaper choice is exactly the same (so given that full-split for all is better than no split for all, I'd keep default-full-split and white list the one to not-split)

File: Configuration/Applications/python/ConfigBuilder.py (hunk 1 above)
Comment (arizzi) [inline Configuration/Applications/python/ConfigBuilder.py]: it is interesting that the python intepreter understands things correctly while github doesn't

**Comment by arizzi** [comment]
OTOH, if one would restrict only to the branches that show some significant difference with full-split / no-split we could reduce the list to ~10 branches (and in that case is probably cheaper, in term of number of python lines, to default no-split and override only a few).

PS: I was expecting more saving when running on less events (1k) that is more "production like" given that production uses fastmerge...  but this is not what I see. I'll do more tests.

File: PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py [hunk 3]
```diff
@@ -94,3 +94,50 @@
...
+cms.untracked.PSet(branch = cms.untracked.string("recoGenJets_slimmedGenJetsAK8__*"),splitLevel=cms.untracked.int32(0)),
+cms.untracked.PSet(branch = cms.untracked.string("recoVertexCompositePtrCandidates_slimmedLambdaVertices__*"),splitLevel=cms.untracked.int32(0)),
+cms.untracked.PSet(branch = cms.untracked.string("GlobalAlgBlkBXVector_gtStage2Digis__*"),splitLevel=cms.untracked.int32(0)),
+cms.untracked.PSet(branch = cms.untracked.string("l1tMuonBXVector_gmtStage2Digis_Muon_*"),splitLevel=cms.untracked.int32(0)),
+cms.untracked.PSet(branch = cms.untracked.string("patPackedCandidates_lostTracks_eleTracks_*"),splitLevel=cms.untracked.int32(0)),
+cms.untracked.PSet(branch = cms.untracked.string("recoConversions_reducedEgamma_reducedSingleLegConversions_*"),splitLevel=cms.untracked.int32(0)),
+cms.untracked.PSet(branch = cms.untracked.string("recoGsfElectronCores_reducedEgamma_reducedGedGsfElectronCores_*"),splitLevel=cms.untracked.int32(0)),
+cms.untracked.PSet(branch = cms.untracked.string("recoPhotonCores_reducedEgamma_reducedGedPhotonCores_*"),splitLevel=cms.untracked.int32(0)),
```
Comment (slava77) [inline PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py]: what about reducedOOT stuff?

File: PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py [hunk 4]
```diff
@@ -94,3 +94,50 @@
...
+cms.untracked.PSet(branch = cms.untracked.string("l1extraL1HFRingss_l1extraParticles__*"),splitLevel=cms.untracked.int32(0)),
+cms.untracked.PSet(branch = cms.untracked.string("l1extraL1EmParticles_l1extraParticles_NonIsolated_*"),splitLevel=cms.untracked.int32(0)),
+cms.untracked.PSet(branch = cms.untracked.string("l1extraL1JetParticles_l1extraParticles_IsoTau_*"),splitLevel=cms.untracked.int32(0)),
+cms.untracked.PSet(branch = cms.untracked.string("l1extraL1JetParticles_l1extraParticles_Forward_*"),splitLevel=cms.untracked.int32(0)),
+cms.untracked.PSet(branch = cms.untracked.string("l1extraL1JetParticles_l1extraParticles_Central_*"),splitLevel=cms.untracked.int32(0)),
+cms.untracked.PSet(branch = cms.untracked.string("l1extraL1EmParticles_l1extraParticles_Isolated_*"),splitLevel=cms.untracked.int32(0)),
+cms.untracked.PSet(branch = cms.untracked.string("l1extraL1MuonParticles_l1extraParticles__*"),splitLevel=cms.untracked.int32(0)),
+cms.untracked.PSet(branch = cms.untracked.string("l1extraL1JetParticles_l1extraParticles_Tau_*"),splitLevel=cms.untracked.int32(0)),
```
Comment (slava77) [inline PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py]: how permissive is this input for wildcards?
Can we have just one ```*_l1extraParticles_*_*``` ?

File: PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py (hunk 2 above)
Comment (slava77) [inline PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py]: some reduced verbosity would be nice, especially for readability.
E.g. 
```python
def split0PSet(name):
   return cms.untracked.PSet(branch = cms.untracked.string(name),splitLevel=cms.untracked.int32(0))

MiniAODOverrideBranchesSplitLevel = cms.untracked.VPSet( [
split0PSet("patMuons_slimmedMuons__*"),
split0PSet("patElectrons_slimmedElectrons__*")
```

**Comment by slava77** [comment]
ping.
@arizzi please follow up on the comments made (6 days ago)

**Comment by arizzi** [comment]
I'm almost off for vacation and have no time for this before I leave. was
there any non stylistic comment?

Il 20 lug 2017 4:07 PM, "Slava Krutelyov" <notifications@github.com> ha
scritto:

**Comment by slava77** [comment]
On 7/20/17 8:04 AM, arizzi wrote:
> I'm almost off for vacation and have no time for this before I leave. was
> there any non stylistic comment?
> 

missing OOT branches is non-stylistic

for the simplification comments,
I find it more of a maintenance issue than style,
in case interface changes
So, the "def split0PSet" and more inclusive use of wildcards applies here.

**Comment by slava77** [comment]
@arizzi 
are you back already to follow up on this PR?
please let me know.
Thank you.

**Comment by arizzi** [comment]
still testing, please do not re-trigger tests yet

**Comment by fwyzard** [comment]
I suspect you need to rebase it to a newer IB ?

**Comment by kpedro88** [comment]
-1
Don't use `cms-merge-topic` in PRs
http://cms-sw.github.io/tutorial-resolve-conflicts.html

See [this comment](https://github.com/cms-sw/cmssw/pull/19771#issuecomment-318504502) for instructions to fix your branch

**Comment by arizzi** [comment]
mhh, I must have committed the merge-topic branch rather than the one with cherry picked commit

**Comment by arizzi** [comment]
forced update with the right branch now. @kpedro88 can you remove your -1

**Comment by kpedro88** [comment]
@arizzi no longer in my category, thanks

**Comment by slava77** [comment]
@cmsbuild please test

@arizzi please clarify if you are done with last round of updates
Thank you.

**Comment by slava77** [comment]
@cmsbuild please test
the last attempt has apparently failed

**Comment by arizzi** [comment]
I still have to run on the 9k stats (the job I sent failed), but if that shows no surprises I'm done with it.

**Comment by slava77** [comment]
@cmsbuild please test

```Build ​timed ​out ​(after ​360 ​minutes). ​Marking ​the ​build ​as ​aborted. ```
@smuzaffar 
I see that there are quite a few comparisons failures .
Is there something going on with the infrastructure, or should we blame it on the huge number of plots with differences due to #19933

**Comment by smuzaffar** [comment]
@slava77 , no it is not infrastructure. compare-root-files-short-matrix jobs are now taking more time to complete. 
May be there are too many files to compare  and here https://github.com/cms-sw/cms-bot/blob/master/run-pr-comparisons#L149 we run all these without control.

**Comment by slava77** [comment]
On 8/1/17 1:22 PM, Malik Shahzad Muzaffar wrote:
> @slava77 <https://github.com/slava77> , no it is not infrastructure.
> compare-root-files-short-matrix jobs are now taking more time to complete.
> May be there are too many files to compare and here
> https://github.com/cms-sw/cms-bot/blob/master/run-pr-comparisons#L149 we
> run all these without control.
> 

the link points to validateJR
I do not expect this to be the reason for a significant slow own.
Are there logs that may suggest this is a problem?

I know from local runs that comparisons for workflows with failing
SiStrip DQM (typically one or two per jenkins run) take about 2 hours each.
I made a PR to address this, at least partially
https://github.com/cms-sw/cms-bot/pull/881

**Comment by smuzaffar** [comment]
this is one run of comparison test (took 3.5 hours) https://meter.cern.ch/public/_plugin/kibana/#dashboard/temp/AV2fkZnFnDFcZnuewq3f which shows  system load reaching 60 (on a 8 core VM). so we are running too many processes in parallel.

**Comment by arizzi** [comment]
anyhow my tests are ok
http://arizzi.web.cern.ch/arizzi/miniaod/splitNewWay.html

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cmsbuild please test

the last few two failed with what seems like some random timeout or similar kill.
In the last attempt the step3 of 136.788 failed with ```ret= 35072 ```
where are these codes defined? @smuzaffar @Dr15Jones

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/19680	2312220
- default split level in miniAOD changed to 0, except for fairly large products (like packedPFCandidates ) which are left in full split mode
- jenkins tests pass and comparisons with baseline show no differences
- local test with 3K events running PAT/miniAOD on ttbar PU35 (an equivalent of wf 10224) confirm the decrease in the event size of about 6% or 3kB/evt. Output module CPU time is unchanged.

Some of the largest reductions (above 0.5% of the output file size)
```
orig      new    delta    delta/fullFile     name
   3071.8 ->      2669.3       -403      -0.80     patMuons_slimmedMuons__PAT.
    237.4 ->         4.3       -233      -0.46     patPhotons_slimmedOOTPhotons__PAT.
   1158.9 ->       691.3       -468      -0.93     patPhotons_slimmedPhotons__PAT.
   1837.5 ->      1395.5       -442      -0.88     patElectrons_slimmedElectrons__PAT.
```

**Comment by slava77** [comment]
@davidlange6 @monttj @franzoni 
analysis and operations signatures are missing for this PR.
Please check and suggest changes or sign.
Thank you.

**Comment by davidlange6** [comment]
merge

**Pull request merged**